### ISSUE TITLE
Capacitor 3 support

### ISSUE THREAD

**Issue by Cannonb4ll** [capacitor-community/barcode-scanner#23]
Great package, awesome you've recently became a community plugin! 🎉

I was wondering if there's Capacitor 3 support coming?

Thanks!

**Comment by tafelnl** [comment]
Yes, it is on the roadmap! Support will likely land before Capacitor 3 is released in stable.

**Comment by agileMikeDev** [comment]
Hi, is capacitor 3 just not supported right now, or does it even not work with the current version 1.2.1?

**Comment by tafelnl** [comment]
I will release a new version that targets Capacitor 3.0 soon.

**Comment by diadal** [comment]
any update on this or any background fix to support cap3

**Comment by kadenn** [comment]
> any update on this or any background fix to support cap3

I think it is possible to update the plugin to cap3 using this guide: https://capacitorjs.com/docs/v3/updating/plugins/3-0

**Comment by diadal** [comment]
> > any update on this or any background fix to support cap3
> 
> I think it is possible to update the plugin to cap3 using this guide: https://capacitorjs.com/docs/v3/updating/plugins/3-0

have tried no luck can you give it a try @kadenn

**Comment by diadal** [comment]
@kadenn  so far https://github.com/diadal/barcode-scanner but not working keep getting this error `Error: "BarcodeScanner.prepare()" is not implemented on ios`

**Comment by diadal** [comment]
@tafelnl with this doc `https://capacitorjs.com/docs/v3/updating/plugins/3-0` i believe there should be quick fix for cap3 can you please check am not good at swift

**Comment by diadal** [comment]
check https://github.com/diadal/barcode-scanner

**Comment by diadal** [comment]
@tafelnl any idea of the time frame

**Comment by tafelnl** [comment]
I will look into it this weekend. Should probably land may 11th

**Comment by Cannonb4ll** [comment]
@diadal Please stop spamming this thread. This is and remains an opensource package and @tafelnl is doing what he can. Wait for him to resolve this issue patiently.

If you are in a rush, fix it yourself and submit a PR.

**Comment by tafelnl** [comment]
I am so sorry guys. My laptop broke down, so I could not work on this project for a while. I am making progress however, so it's coming real soon!

**Comment by diadal** [comment]
sorry to hear @tafelnl since @Cannonb4ll believed I was spamming I try to work on a supported version for cap3 to do what I needed for my project am working on, this only works on ios the code were different from yours so I cant submit PR you can check https://github.com/diadal/barcode-scanner3 it may be of help. 

To have both android & iOs under the same package it better 

maybe you can add a Sponsorship button we user can be of help

**Comment by hussainarthuna** [comment]
> I am so sorry guys. My laptop broke down, so I could not work on this project for a while. I am making progress however, so it's coming real soon!

Hey looking forward for a new update as the V3 has been released and I waiting on this package to be updated so that I can upgrade my Application. Thank you.

**Comment by e109923** [comment]
Every day I check this page... I need to updateeee :)

**Comment by tafelnl** [comment]
Please be patient guys. I just got a new machine to work on and installed everything. So I am up and running again. I just have to finish off a little bit of work. I have given this issue a high priority on my to do list.

New ETA: 23th of may.

**Comment by quangpdt** [comment]
Thanks @tafelnl for the good news. Btw, I made a fork to upgrade your plugin for compatibility with Capacitor v3. Unfortunately, it's only for iOS though (We're only working with iOS platform), but you can use that for reference, I would not create PR for this 

https://github.com/mobiliarmad/barcode-scanner

**Comment by tafelnl** [comment]
I have very good news for all of you! I just released an alpha version with full support for Capacitor 3: https://github.com/capacitor-community/barcode-scanner/releases/tag/2.0.0-alpha.0

Would you please try it out and report back to me in this issue? That way we can validate that no bugs / issues are present. If nobody experiences any problems, I will release it to production ASAP.

**Comment by bolchowka** [comment]
The updated version doesn't seem to work for me.

Here's my module version from `package.json`:
```
"@capacitor-community/barcode-scanner": "^2.0.0-alpha.0",
```

Here's my import:
```ts
import { BarcodeScanner } from "@capacitor-community/barcode-scanner";
```

And here's my code:
```ts
BarcodeScanner.startScan({})
  .then(value => console.log(value))
  .catch(e => console.warn(e));
```

The promise neither resolves nor rejects. It hangs indefinitely. No UI is displayed to start scanning. `checkPermission` method returns `{neverAsked: true}`.

**Comment by bolchowka** [comment]
I've done some debug in `Plugin.swift` and I can't tell why the scanner doesn't appear. Camera permissions are fine (`{"granted":true}`) and the plugin code reaches the final place where the scanning should start:
```swift
DispatchQueue.main.async {
    self.metaOutput!.metadataObjectTypes = self.targetedFormats
    self.captureSession!.startRunning()
    print("start scanning") // <--- I can see that in my XCode console
}
```
However, nothing happens.

I suspect it might be connected with the opacity of my webview. Quoting README:
> Because of the fact that the Scanner View will be rendered behind the WebView, you will have to call hideBackground() to make the WebView and the <html> element transparent. Every other element that needs transparency, you will have to handle yourself.

I do use `BarcodeScanner.hideBackground()`, but maybe there's anything else that needs to be done?

**Comment by diadal** [comment]
@tafelnl really need to check 
https://github.com/mobiliarmad/barcode-scanner
and 
https://github.com/diadal/barcode-scanner3

the way ios13 above and cap3 handle thing really different also I will guest you create a different branch for cap3

**Comment by tafelnl** [comment]
@bolchowka You could check out these issues: https://github.com/capacitor-community/barcode-scanner/issues/26 and https://github.com/capacitor-community/barcode-scanner/issues/7

Those issues provide some more information on hiding the background

**Comment by tafelnl** [comment]
@diadal I am really not sure what you mean. I test on iOS 12.x and 14.x and both verified working. I already created different branches.

Please read the upgrade guide here: https://github.com/capacitor-community/barcode-scanner/releases/tag/2.0.0-alpha.0

**Comment by bolchowka** [comment]
> @bolchowka You could check out these issues: #26 and #7
> 
> Those issues provide some more information on hiding the background

That was it. Playing with opacity on `html` and `body` elements helped. Thank you!

**Comment by phoenix503** [comment]
This issue is coming to me 
`Compiling for iOS 11.0, but module 'Capacitor' has a minimum deployment target of iOS 12.0:`
@tafelnl  please check
Have you fixed it ?

**Comment by tafelnl** [comment]
@phoenix503 Please read through this issue. If you are using or want to use Capacitor 3.0 this is migration guide is a guide start: https://github.com/capacitor-community/barcode-scanner/releases/tag/2.0.0-alpha.0

If you want to use Capacitor 2.x please install v1.x of this plugin

**Comment by tafelnl** [comment]
`alpha` has over 150 installs on NPM and problems seem to be detected so far. So I think we can be safe to publish this to production by 2th of June.

**Comment by JZubero** [comment]
@tafelnl Great work on Capacitor 3 support :clap: Works pretty smooth for us so far (tested with Android + iOS).

One thing is odd, though. The `scan` method has one mandatory parameter (object), which may be empty since the only property of the parameter object (`targetedFormats`) is optional. May I ask what the intention behind the mandatory but possible empty param object is?

```typescript
export interface ScanOptions {
    targetedFormats?: [SupportedFormat];
}
```

Furthermore, I do not really get the value typing for `targetedFormats` :thinking: With `SupportedFormat` being an enum with all possible code formats, I seem to be only able to pass one target format. Otherwise my linter is unhappy or I do not get any scan result at all.

### Unhappy linter

![unhappy-linter](https://user-images.githubusercontent.com/13751426/120197932-dbaa1c00-c221-11eb-85bd-d38dcff559b2.png)


### No scan result

![no-result](https://user-images.githubusercontent.com/13751426/120197976-e6fd4780-c221-11eb-93e0-24bdf3f15728.png)


### Works
...but only with one single target format specified.

![works](https://user-images.githubusercontent.com/13751426/120198027-f5e3fa00-c221-11eb-951f-799b708284de.png)

**Comment by desbrowne** [comment]
@JZubero I also noticed the issue with the formats.

@tafelnl the targetedFormats field in definitions.ts is type `[SupportedFormat]`, which means a one element array. It seems that the type should be `SupportedFormat[]`?

**Comment by tafelnl** [comment]
@JZubero @desbrowne You are correct. I just fixed the types in `v2.0.0-alpha.1` and released it to npm

**Comment by elvisgraho** [comment]
Hi, im using ```v2.0.0-alpha.1``` in my ionic app (capacitor 3). It works, but I have the opacity issues (the scanner does not appear), what is the way to fix it?

**Comment by tafelnl** [comment]
Please take a look at the docs here: https://github.com/capacitor-community/barcode-scanner#the-scanner-view-does-not-show-up

**Comment by tafelnl** [comment]
Just released `v2.0.0`. More info: https://github.com/capacitor-community/barcode-scanner/releases/tag/v2.0.0

Closing this issue now.
